src/client/binary_client.cpp

```diff
@@ -822,39 +824,64 @@ namespace
+      if(responseHeader.Chunk == CHT_SINGLE)
+      {
+        parseMessage(dataSize, id);			  
+        firstMsgParsed = false;
+
+        std::unique_lock<std::mutex> lock(Mutex);
+        CallbackMap::const_iterator callbackIt = Callbacks.find(header.RequestHandle);
+        if (callbackIt == Callbacks.end())
+        {
+          std::cout << "binary_client| No callback found for message with id: " << id << " and handle " << header.RequestHandle << std::endl;
+          return;
+        }
+        callbackIt->second(std::move(messageBuffer), std::move(header));
+        messageBuffer.clear();
+        Callbacks.erase(callbackIt);
+			}
+      else if(responseHeader.Chunk == CHT_INTERMEDIATE)
+      {
+        parseMessage(dataSize, id);
+        firstMsgParsed = true;		  
+      }	  
+    }
+
+    ResponseHeader parseMessage(std::size_t &dataSize, NodeId &id)
+    {
       std::vector<char> buffer(dataSize);
       BufferInputChannel bufferInput(buffer);
       Binary::RawBuffer raw(&buffer[0], dataSize);
+
       Stream >> raw;
 
-      IStreamBinary in(bufferInput);
-      NodeId id;
-      in >> id;
-      ResponseHeader header;
-      in >> header;
-      if ( Debug )std::cout << "binary_client| Got response id: " << id << " and handle " << header.RequestHandle<< std::endl;
+      if(!firstMsgParsed)
+      {
+        IStreamBinary in(bufferInput);
+        in >> id;
+        in >> header;
 
-      if (header.ServiceResult != StatusCode::Good) {
-        std::cout << "binary_client| Received a response from server with error status: " << OpcUa::ToString(header.ServiceResult) <<  std::endl;
-      }
+        if ( Debug )std::cout << "binary_client| Got response id: " << id << " and handle " << header.RequestHandle<< std::endl;
 
-      if (id == SERVICE_FAULT)
-      {
-        std::cerr << std::endl;
-        std::cerr << "Receive ServiceFault from Server with StatusCode " << OpcUa::ToString(header.ServiceResult) << std::endl;
-        std::cerr << std::endl;
+        if (header.ServiceResult != StatusCode::Good) {
+          std::cout << "binary_client| Received a response from server with error status: " << OpcUa::ToString(header.ServiceResult) <<  std::endl;
+        }
+
+        if (id == SERVICE_FAULT)
+        {
+          std::cerr << std::endl;
+          std::cerr << "Receive ServiceFault from Server with StatusCode " << OpcUa::ToString(header.ServiceResult) << std::endl;
+          std::cerr << std::endl;
+        }
+
+        messageBuffer.insert(messageBuffer.end(), buffer.begin(), buffer.end());
+        return header;
       }
-      std::unique_lock<std::mutex> lock(Mutex);
-      CallbackMap::const_iterator callbackIt = Callbacks.find(header.RequestHandle);
-      if (callbackIt == Callbacks.end())
+      else
       {
-        std::cout << "binary_client| No callback found for message with id: " << id << " and handle " << header.RequestHandle << std::endl;
-        return;
```

Comment:
I get a warning, because the return is missing: src/client/binary_client.cpp:886:5: warning: control reaches end of non-void function